# 패키지 참조 및 데이터 가져오기

In [35]:
from hossam import load_data
from matplotlib import pyplot as plt
from matplotlib import font_manager as fm
import seaborn as sb
import numpy as np
import pandas as pd
from pandas import DataFrame
import statsmodels.api as sm
from statsmodels.stats.diagnostic import linear_reset
from scipy.stats import zscore, pearsonr, spearmanr, probplot, shapiro, kstest, anderson

from statsmodels.formula.api import ols
from statsmodels.stats.api import het_breuschpagan
from statsmodels.stats.stattools import durbin_watson

from statsmodels.stats.outliers_influence import variance_inflation_factor

In [36]:
origin=load_data('diamonds')
print('\n===== 데이터 크기 확인 =====')
print(f'데이터셋 크기 : {origin.shape}')
print(f'행 개수 : {origin.shape[0]}개')
print(f'열 개수 : {origin.shape[1]}개')
print('\n===== 타입 확인 =====')
print(origin.info())
origin.head()

[data] https://data.hossam.kr/data/kaggle/diamonds.xlsx
[desc] 다이아몬드를 커팅, 색상, 투명도, 가격 및 기타 속성 조사한 데이터 셋 (출처: https://www.kaggle.com/datasets/shivam2503/diamonds)

field    description
-------  -------------------------------------------------------------------
price    다이아몬드 가격 (USD, $326 ~ $18,823)
carat    중량 (0.2~5.01)
cut      컷 품질 (Fair, Good, Very Good, Premium, Ideal)
color    색상 등급 - J (worst) to D (best)
clarity  투명도 등급 (I1 (worst), SI2, SI1, VS2, VS1, VVS2, VVS1, IF (best))
x        길이 mm (0~10.74)
y        너비 mm (0--58.9)
z        두께 mm (0--31.8)
depth    비율 정보 = z / mean(x, y) = 2 * z / (x + y) (43--79)
table    다이아몬드의 가장 넓은 지점에 비해 상단(테이블)의 너비 (43--95)


===== 데이터 크기 확인 =====
데이터셋 크기 : (53940, 10)
행 개수 : 53940개
열 개수 : 10개

===== 타입 확인 =====
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 53940 entries, 0 to 53939
Data columns (total 10 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   price    53940 non-null  int64  
 1   carat    

price  carat      cut color clarity     x     y     z  depth  table
0    326   0.23    Ideal     E     SI2  3.95  3.98  2.43   61.5   55.0
1    326   0.21  Premium     E     SI1  3.89  3.84  2.31   59.8   61.0
2    327   0.23     Good     E     VS1  4.05  4.07  2.31   56.9   65.0
3    334   0.29  Premium     I     VS2  4.20  4.23  2.63   62.4   58.0
4    335   0.31     Good     J     SI2  4.34  4.35  2.75   63.3   58.0

# 전처리 작업

In [37]:
# 결측치 확인

origin.isna().sum()

price      0
carat      0
cut        0
color      0
clarity    0
x          0
y          0
z          0
depth      0
table      0
dtype: int64

> 값이 Nan인 결측치는 존재하지 않음

In [38]:
# 중복 데이터 확인

origin.duplicated().sum()

np.int64(146)

> 146개의 중복 행이 검사됨. 이를 제거 후 이어서 분석 진행.

In [39]:
df1=origin.drop_duplicates()

In [40]:
# x, y, z 값이 0 이하인 데이터 비율 확인

print('x, y, z 값이 0 이하인 데이터 수 :',df1[(df1['x']<=0) | (df1['y']<=0) | (df1['z']<=0)]['price'].count())
print('x, y, z 값이 0 이하인 데이터 비율 :',(df1[(df1['x']<=0) | (df1['y']<=0) | (df1['z']<=0)]['price'].count())/len(df1)*100)

x, y, z 값이 0 이하인 데이터 수 : 19
x, y, z 값이 0 이하인 데이터 비율 : 0.03531992415511024


> 다이아몬드의 크기를 결정하는 x, y, z가 0 이하인 데이터는 결측치로 판단

> 결측률이 5% 미만이기에 결측치를 제거 후 데이터 분석 진행

In [41]:
df=df1.drop(index=df1[(df1['x']==0) | (df1['y']==0) | (df1['z']==0)].index)

In [42]:
# 원본 데이터와 전처리 이후 데이터 수 비교

print('원본 데이터 수 :', len(origin))
print('전처리 이후 데이터 수 :', len(df))
print('=== 데이터 %d건 삭제 ==='%(len(origin)-len(df)))

원본 데이터 수 : 53940
전처리 이후 데이터 수 : 53775
=== 데이터 165건 삭제 ===


# 사분위수 데이터 분리

In [43]:
q1, q2, q3=df['price'].quantile([0.25, 0.5, 0.75])
print(q1)
print(q2)
print(q3)

951.0
2401.0
5324.0


In [44]:
df_q1=df[df['price']<949]
df_q2=df[(df['price']>=949) & (df['price']<2401)]
df_q3=df[(df['price']>=2401) & (df['price']<5323.25)]
df_q4=df[df['price']>=5323.25]
display(df_q1)
display(df_q2)
display(df_q3)
display(df_q4)

price  carat        cut color clarity     x     y     z  depth  table
0        326   0.23      Ideal     E     SI2  3.95  3.98  2.43   61.5   55.0
1        326   0.21    Premium     E     SI1  3.89  3.84  2.31   59.8   61.0
2        327   0.23       Good     E     VS1  4.05  4.07  2.31   56.9   65.0
3        334   0.29    Premium     I     VS2  4.20  4.23  2.63   62.4   58.0
4        335   0.31       Good     J     SI2  4.34  4.35  2.75   63.3   58.0
...      ...    ...        ...   ...     ...   ...   ...   ...    ...    ...
53635    552   0.30      Ideal     I    VVS1  4.30  4.33  2.68   62.1   56.0
53636    552   0.30      Ideal     D     SI1  4.28  4.32  2.65   61.6   57.0
53637    552   0.30  Very Good     D     SI1  4.26  4.28  2.67   62.5   57.0
53638    552   0.30      Ideal     D     SI1  4.29  4.33  2.63   61.0   57.0
53639    552   0.30      Ideal     D     SI1  4.35  4.37  2.66   61.0   56.0

[13414 rows x 10 columns]

price  carat        cut color clarity     x     y     z  depth  table
36651    949   0.35    Premium     J     VS1  5.67  5.64  3.53   62.4   58.0
36653    949   0.40    Premium     E     VS2  4.79  4.68  2.90   61.2   60.0
36654    949   0.30  Very Good     E      IF  4.29  4.32  2.62   60.9   59.0
36655    949   0.41       Good     D    VVS2  4.71  4.73  2.99   63.3   59.0
36656    949   0.30      Ideal     E    VVS2  4.35  4.37  2.65   60.8   56.0
...      ...    ...        ...   ...     ...   ...   ...   ...    ...    ...
51704   2400   0.70  Very Good     F     VS2  5.72  5.78  3.52   61.2   56.0
51705   2400   0.70  Very Good     F     VS2  5.79  5.85  3.43   58.9   57.0
51706   2400   0.58  Very Good     D    VVS2  5.28  5.35  3.35   63.0   56.0
51707   2400   0.71      Ideal     G     SI1  5.77  5.79  3.54   61.2   58.0
51708   2400   0.80       Fair     G     SI2  5.76  5.73  3.85   67.1   56.0

[13460 rows x 10 columns]

price  carat        cut color clarity     x     y     z  depth  table
90      2757   0.70      Ideal     E     SI1  5.70  5.72  3.57   62.5   57.0
91      2757   0.86       Fair     E     SI2  6.45  6.33  3.52   55.1   69.0
92      2757   0.70      Ideal     G     VS2  5.70  5.67  3.50   61.6   56.0
93      2759   0.71  Very Good     E     VS2  5.68  5.73  3.56   62.4   57.0
94      2759   0.78  Very Good     G     SI2  5.81  5.85  3.72   63.8   56.0
...      ...    ...        ...   ...     ...   ...   ...   ...    ...    ...
53935   2757   0.72      Ideal     D     SI1  5.75  5.76  3.50   60.8   57.0
53936   2757   0.72       Good     D     SI1  5.69  5.75  3.61   63.1   55.0
53937   2757   0.70  Very Good     D     SI1  5.66  5.68  3.56   62.8   60.0
53938   2757   0.86    Premium     H     SI2  6.15  6.12  3.74   61.0   58.0
53939   2757   0.75      Ideal     D     SI2  5.83  5.87  3.64   62.2   55.0

[13454 rows x 10 columns]

price  carat        cut color clarity     x     y     z  depth  table
12760   5324   1.21    Premium     H     SI2  6.75  6.78  4.24   62.7   59.0
12761   5324   1.21  Very Good     H     SI2  6.70  6.75  4.21   62.6   61.0
12762   5324   1.21  Very Good     I     SI1  6.76  6.92  4.25   62.1   56.0
12763   5324   1.21      Ideal     J     VS1  6.81  6.86  4.26   62.3   55.0
12764   5324   0.90  Very Good     F     VS1  6.18  6.23  3.80   61.2   56.0
...      ...    ...        ...   ...     ...   ...   ...   ...    ...    ...
27745  18803   2.00  Very Good     H     SI1  7.95  8.00  5.01   62.8   57.0
27746  18804   2.07      Ideal     G     SI2  8.20  8.13  5.11   62.5   55.0
27747  18806   1.51      Ideal     G      IF  7.37  7.41  4.56   61.7   55.0
27748  18818   2.00  Very Good     G     SI1  7.90  7.97  5.04   63.5   56.0
27749  18823   2.29    Premium     I     VS2  8.50  8.47  5.16   60.8   60.0

[13447 rows x 10 columns]

# 사분위수 선형회귀 분석

In [45]:
def ols_report(tbl, data):
    xnames=[n for n in fit.model.exog_names if n!='const']
    indi_mydata1=data.filter(xnames)
    variables=[]
    for i, v in enumerate(tbl.tables[1].data):
        name=v[0].strip()
        if name not in xnames:
            continue
        j=list(indi_mydata1.columns).index(name)
        vif=variance_inflation_factor(indi_mydata1,j)

        p=float(v[4].strip())
        stars=lambda p: (
            '***' if p<0.001 else
            '**' if p<0.01 else
            '*' if p<0.05 else
            ''
        )

        variables.append({
            '종속변수':yname,
            '독립변수':name,
            'B':v[1].strip(),
            '표준오차':v[2].strip(),
            'Beta':float(fit.params[name])*(data[name].std(ddof=1)/data[yname].std(ddof=1)),
            't':'%s%s'%(v[3].strip(), stars(p)),
            'p-value':p,
            '공차':1/vif,
            'vif':vif
        })

    rdf=DataFrame(variables)

    if rdf.empty:
        return None, "데이터 없음", "모델 유의하지 않음", []

    result_dict={}
    for i in [0,2]:
        for item in tbl.tables[i].data:
            n=len(item)
            for j in range(0, n, 2):
                key=item[j].strip()[:-1]
                value=item[j+1].strip()
                if not key or not value:
                    continue
                result_dict[key]=value
    
    result_report=f'R({result_dict['R-squared']}), R²({result_dict['Adj. R-squared']}), F({result_dict['F-statistic']}), 유의확률({result_dict['Prob (F-statistic)']}), Durbin-Watson({result_dict['Durbin-Watson']})'

    tpl='%s에 대하여 %s로 예측하는 회귀분석을 실시한 결과, 이 회귀모형은 통계적으로 %s(F(%s,%s)=%s, p %s 0.05).'
    model_report=tpl%(
        rdf['종속변수'][0],
        ','.join(list(rdf['독립변수'])),
        '유의하다' if float(result_dict['Prob (F-statistic)'])<=0.05 else '유의하지 않다',
        result_dict['Df Model'],
        result_dict['Df Residuals'],
        result_dict['F-statistic'],
        '<=' if float(result_dict['Prob (F-statistic)'])<=0.05 else '>'
    )

    variable_reports=[]
    s='%s의 회귀계수는 %s(p %s 0.05)로, %s에 대하여 %s 예측변인인 것으로 나타났다.'

    for i in rdf.index:
        row=rdf.iloc[i]
        variable_reports.append(s%(row['독립변수'],
                                   row['B'],
                                   '<=' if float(row['p-value'])<0.05 else '>',
                                   row['종속변수'],
                                   '유의미한' if float(row['p-value'])<0.05 else '유의하지 않은'))
        
    return rdf, result_report, model_report, variable_reports

In [46]:
print('='*50)
print('🎓 1사분위수 선형회귀 분석')
print('='*50)
mydata=df_q1
yname='price'
# mydata1=mydata.copy()
mydata1=mydata.drop(['cut','color','clarity'], axis=1)
x=mydata1.drop(yname, axis=1)
y=mydata1[yname]
x_input=sm.add_constant(x)
fit=sm.OLS(y, x_input).fit()
tbl1=fit.summary()
rdf, r, m, v=ols_report(tbl1, mydata)
display(rdf)

print('='*50)
print('🎓 2사분위수 선형회귀 분석')
print('='*50)
mydata=df_q2
yname='price'
# mydata1=mydata.copy()
mydata1=mydata.drop(['cut','color','clarity'], axis=1)
x=mydata1.drop(yname, axis=1)
y=mydata1[yname]
x_input=sm.add_constant(x)
fit=sm.OLS(y, x_input).fit()
tbl1=fit.summary()
rdf, r, m, v=ols_report(tbl1, mydata)
display(rdf)

print('='*50)
print('🎓 3사분위수 선형회귀 분석')
print('='*50)
mydata=df_q3
yname='price'
# mydata1=mydata.copy()
mydata1=mydata.drop(['cut','color','clarity'], axis=1)
x=mydata1.drop(yname, axis=1)
y=mydata1[yname]
x_input=sm.add_constant(x)
fit=sm.OLS(y, x_input).fit()
tbl1=fit.summary()
rdf, r, m, v=ols_report(tbl1, mydata)
display(rdf)

print('='*50)
print('🎓 4사분위수 선형회귀 분석')
print('='*50)
mydata=df_q4
yname='price'
# mydata1=mydata.copy()
mydata1=mydata.drop(['cut','color','clarity'], axis=1)
x=mydata1.drop(yname, axis=1)
y=mydata1[yname]
x_input=sm.add_constant(x)
fit=sm.OLS(y, x_input).fit()
tbl1=fit.summary()
rdf, r, m, v=ols_report(tbl1, mydata)
display(rdf)

🎓 1사분위수 선형회귀 분석


종속변수   독립변수           B     표준오차      Beta           t  p-value        공차  \
0  price  carat    418.2261  177.342  0.139142      2.358*    0.018  0.000535   
1  price      x    712.1382   36.146  1.049620   19.702***    0.000  0.000079   
2  price      y  -1268.6392   54.634 -1.868168  -23.221***    0.000  0.000072   
3  price      z   1269.4950   97.301  1.191316   13.047***    0.000  0.000037   
4  price  depth    -64.9594    4.759 -0.501661  -13.650***    0.000  0.000099   
5  price  table     -7.9740    0.649 -0.111636  -12.286***    0.000  0.001115   

            vif  
0   1870.038224  
1  12623.340831  
2  13923.200928  
3  26980.367784  
4  10057.396023  
5    896.727561

🎓 2사분위수 선형회귀 분석


종속변수   독립변수         B     표준오차      Beta           t  p-value        공차  \
0  price  carat  748.2769  137.160  0.210778    5.455***    0.000  0.004683   
1  price      x  591.0668   43.322  0.530960   13.644***    0.000  0.000421   
2  price      y   44.1691   10.236  0.045834    4.315***    0.000  0.002051   
3  price      z   20.9140    9.663  0.016791      2.164*    0.030  0.006027   
4  price  depth   -6.3604    2.578 -0.020486     -2.467*    0.014  0.001481   
5  price  table  -14.9387    1.181 -0.076082  -12.648***    0.000  0.001713   

           vif  
0   213.560376  
1  2373.447083  
2   487.504219  
3   165.919778  
4   675.234076  
5   583.624729

🎓 3사분위수 선형회귀 분석


종속변수   독립변수          B     표준오차      Beta          t  p-value        공차  \
0  price  carat   -86.9371  242.356 -0.017683     -0.359    0.720  0.000613   
1  price      x  -865.3313  116.747 -0.404999  -7.412***    0.000  0.000079   
2  price      y  1966.7515  104.547  0.894859  18.812***    0.000  0.000087   
3  price      z  1002.2903  235.376  0.296078   4.258***    0.000  0.000045   
4  price  depth   -10.8179   13.550 -0.020840     -0.798    0.425  0.000126   
5  price  table    -7.7796    2.350 -0.021145   -3.310**    0.001  0.001353   

            vif  
0   1631.709194  
1  12680.327284  
2  11494.994161  
3  22277.843925  
4   7936.615796  
5    739.216380

🎓 4사분위수 선형회귀 분석


종속변수   독립변수          B     표준오차      Beta           t  p-value        공차  \
0  price  carat  3378.0016  407.806  0.359631    8.283***    0.000  0.002149   
1  price      x   982.3228  295.901  0.168462     3.320**    0.001  0.000428   
2  price      y   -71.9458   54.493 -0.015013      -1.320    0.187  0.003056   
3  price      z  2307.8508  400.436  0.247387    5.763***    0.000  0.000169   
4  price  depth  -276.9024   33.061 -0.109719   -8.376***    0.000  0.000418   
5  price  table  -126.4117   10.537 -0.073943  -11.997***    0.000  0.001398   

           vif  
0   465.322921  
1  2335.980463  
2   327.200469  
3  5906.367965  
4  2391.263476  
5   715.392752